File: bridges/python/swufe_bridge/wrd_codec.py
"""SWUFE / 网瑞达 (WRD) WebVPN URL codec.

Confirmed against a live webvpn.swufe.edu.cn address-bar URL (2026-09-20):

  AES-128-CFB, segment_size=128 (CFB128)
  key = iv = b"wrdvpnisthebest!"
  form:
    https://WEBVPN_HOST/{scheme}[-{port}]/{iv_hex}{encrypted_host_hex}{path}?{query}

Only the hostname is encrypted; path and query stay plaintext.

The AES primitive comes from ``cryptography`` (a transitive dependency of
mitmproxy); the archived prototype used pycryptodome, which is not a
dependency of this project.
"""

from __future__ import annotations

import argparse
import binascii
import re
import sys
from typing import Optional
from urllib.parse import urlparse, urlunparse

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

try:  # cryptography >= 43 (CFB lives in `decrepit` from 49 on)
    from cryptography.hazmat.decrepit.ciphers.modes import CFB
except ImportError:  # pragma: no cover - cryptography < 43
    from cryptography.hazmat.primitives.ciphers.modes import CFB

DEFAULT_KEY = DEFAULT_IV = "wrdvpnisthebest!"
DEFAULT_WEBVPN_HOST = "webvpn.swufe.edu.cn"

_SCHEME_RE = re.compile(r"^(?P<scheme>https?|http|https)(?:-(?P<port>\d+))?$")


class WrdCodecError(ValueError):
    """Raised for malformed keys/IVs, hosts, tokens and WRD URLs."""


def _key_bytes(value: str | bytes, what: str) -> bytes:
    raw = value.encode("utf-8") if isinstance(value, str) else bytes(value)
    if len(raw) != 16:
        raise WrdCodecError(f"WRD AES-128 {what} must be 16 bytes")
    return raw


class WrdCodec:
    """Encrypt / decrypt WRD WebVPN host segments and full URLs."""

    def __init__(
        self,
        key: str | bytes = DEFAULT_KEY,
        iv: str | bytes = DEFAULT_IV,
        webvpn_host: str = DEFAULT_WEBVPN_HOST,
    ) -> None:
        self.key = _key_bytes(key, "key")
        self.iv = _key_bytes(iv, "IV")
        self.webvpn_host = webvpn_host.rstrip("/")

    def _cipher(self, iv: bytes):
        return Cipher(algorithms.AES(self.key), CFB(iv))

    def encrypt_host(self, host: str) -> str:
        """Return iv_hex + ciphertext_hex for a hostname (no port)."""
        encryptor = self._cipher(self.iv).encryptor()
        ct = encryptor.update(host.encode("utf-8")) + encryptor.finalize()
        return binascii.hexlify(self.iv).decode("ascii") + binascii.hexlify(ct).decode("ascii")

    def decrypt_host(self, token: str) -> str:
        """Decrypt iv_hex+ciphertext_hex back to a hostname."""
        if len(token) < 34 or len(token) % 2:
            raise WrdCodecError("host token too short or not hex")
        iv_hex, ct_hex = token[:32], token[32:]
        try:
            iv = binascii.unhexlify(iv_hex)
            ct = binascii.unhexlify(ct_hex)
        except (binascii.Error, ValueError) as exc:
            raise WrdCodecError(f"host token too short or not hex: {exc}") from exc
        # Prefer the IV carried in the URL (matches portal.js); fall back to the configured IV.
        use_iv = iv if len(iv) == 16 else self.iv
        decryptor = self._cipher(use_iv).decryptor()
        pt = decryptor.update(ct) + decryptor.finalize()
        try:
            return pt.decode("utf-8").rstrip("\x00")
        except UnicodeDecodeError as exc:
            raise WrdCodecError(f"host token does not decrypt to a hostname: {exc}") from exc

    def encode_url(self, ordinary_url: str, webvpn_base: str | None = None) -> str:
        """Ordinary https://host/path → WebVPN URL."""
        parsed = urlparse(ordinary_url)
        if parsed.scheme not in ("http", "https"):
            raise WrdCodecError(f"unsupported scheme: {parsed.scheme!r}")
        if not parsed.hostname:
            raise WrdCodecError("missing hostname")

        scheme = parsed.scheme
        port = parsed.port
        default_port = 443 if scheme == "https" else 80
        if port is None or port == default_port:
            scheme_token = scheme
        else:
            scheme_token = f"{scheme}-{port}"

        token = self.encrypt_host(parsed.hostname)
        path = parsed.path or "/"
        if not path.startswith("/"):
            path = "/" + path

        base = (webvpn_base or f"https://{self.webvpn_host}").rstrip("/")
        out = f"{base}/{scheme_token}/{token}{path}"
        if parsed.query:
            out += f"?{parsed.query}"
        if parsed.fragment:
            out += f"#{parsed.fragment}"
        return out

    def decode_url(self, webvpn_url: str) -> str:
        """WebVPN URL → ordinary URL."""
        parsed = urlparse(webvpn_url)
        # path: /{scheme_token}/{token}{rest_path}
        raw = parsed.path
        if not raw.startswith("/"):
            raise WrdCodecError("invalid WebVPN path")
        pieces = raw.lstrip("/").split("/", 2)
        if len(pieces) < 2:
            raise WrdCodecError("WebVPN path missing scheme or host token")
        scheme_token, token = pieces[0], pieces[1]
        rest = "/" + pieces[2] if len(pieces) > 2 else "/"

        m = _SCHEME_RE.match(scheme_token)
        if not m:
            raise WrdCodecError(f"bad scheme token: {scheme_token!r}")
        scheme = m.group("scheme")
        port_s = m.group("port")
        port = int(port_s) if port_s else None

        host = self.decrypt_host(token)
        netloc = host if port is None else f"{host}:{port}"
        return urlunparse((scheme, netloc, rest, "", parsed.query, parsed.fragment))


def main(argv: Optional[list[str]] = None) -> int:
    parser = argparse.ArgumentParser(description="SWUFE WRD WebVPN URL codec")
    parser.add_argument("--key", default=DEFAULT_KEY, help="16-byte AES key string")
    parser.add_argument("--iv", default=DEFAULT_IV, help="16-byte AES IV string")
    parser.add_argument("--webvpn-host", default=DEFAULT_WEBVPN_HOST)
    sub = parser.add_subparsers(dest="cmd", required=True)

    p_enc = sub.add_parser("encode", help="ordinary URL → WebVPN URL")
    p_enc.add_argument("url")

    p_dec = sub.add_parser("decode", help="WebVPN URL → ordinary URL")
    p_dec.add_argument("url")

    args = parser.parse_args(argv)
    try:
        codec = WrdCodec(key=args.key, iv=args.iv, webvpn_host=args.webvpn_host)
        if args.cmd == "encode":
            print(codec.encode_url(args.url))
        else:
            print(codec.decode_url(args.url))
    except WrdCodecError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/webvpn-core-js/src/codec/wrd-codec.ts
import { aesCfb128Decrypt, aesCfb128Encrypt } from "./aes-cfb128.ts";
import { CodecError, type CodecErrorCode } from "../errors.ts";

export const DEFAULT_KEY = "wrdvpnisthebest!";
export const DEFAULT_IV = "wrdvpnisthebest!";
export const DEFAULT_WEBVPN_HOST = "webvpn.swufe.edu.cn";

export interface WrdCodecOptions {
  gatewayBase: string;
  key: Uint8Array;
  iv: Uint8Array;
}

const SCHEME_RE = /^(?<scheme>https?|http|https)(?:-(?<port>\d+))?$/;

export class WrdCodec {
  readonly key: Uint8Array;
  readonly iv: Uint8Array;
  readonly webvpnHost: string;

  constructor(
    key: string | Uint8Array = DEFAULT_KEY,
    iv: string | Uint8Array = DEFAULT_IV,
    webvpnHost: string = DEFAULT_WEBVPN_HOST,
  ) {
    this.key = keyBytes(key, "INVALID_KEY");
    this.iv = keyBytes(iv, "INVALID_IV");
    this.webvpnHost = webvpnHost.replace(/\/+$/, "");
  }

  static fromOptions(options: WrdCodecOptions): WrdCodec {
    const host = gatewayHost(options.gatewayBase);
    return new WrdCodec(options.key, options.iv, host);
  }

  encryptHost(host: string): string {
    const ct = aesCfb128Encrypt(this.key, this.iv, utf8(host));
    return toHex(this.iv) + toHex(ct);
  }

  decryptHost(token: string): string {
    if (token.length < 34 || token.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(token)) {
      throw new CodecError("INVALID_TOKEN");
    }
    const iv = fromHex(token.slice(0, 32));
    const ct = fromHex(token.slice(32));
    if (!iv || !ct || iv.length !== 16) {
      throw new CodecError("INVALID_TOKEN");
    }
    const pt = aesCfb128Decrypt(this.key, iv, ct);
    const text = utf8Decode(pt).replace(/\0+$/, "");
    if (text.length === 0 || text.includes("\uFFFD")) {
      throw new CodecError("DECRYPT_FAILED");
    }
    return text;
  }

  encodeUrl(ordinaryUrl: string, webvpnBase?: string): string {
    if (/^[a-z][a-z0-9+.-]*:\/\/(?:\/|$)/i.test(ordinaryUrl)) {
      throw new CodecError("INVALID_URL");
    }
    const parsed = parseUrl(ordinaryUrl);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
      throw new CodecError("UNSUPPORTED_SCHEME");
    }
    if (!parsed.hostname) {
      throw new CodecError("INVALID_URL");
    }
    const scheme = parsed.protocol === "https:" ? "https" : "http";
    const port = parsed.port === "" ? null : Number(parsed.port);
    const defaultPort = scheme === "https" ? 443 : 80;
    const schemeToken = port === null || port === defaultPort ? scheme : `${scheme}-${port}`;
    const token = this.encryptHost(parsed.hostname);
    const path = parsed.pathname.startsWith("/") ? parsed.pathname : `/${parsed.pathname}`;
    const base = (webvpnBase ?? `https://${this.webvpnHost}`).replace(/\/+$/, "");
    let out = `${base}/${schemeToken}/${token}${path}`;
    if (parsed.search) {
      out += parsed.search;
    }
    if (parsed.hash) {
      out += parsed.hash;
    }
    return out;
  }

  decodeUrl(webvpnUrl: string): string {
    let parsed: URL;
    try {
      parsed = new URL(webvpnUrl);
    } catch {
      throw new CodecError("INVALID_URL");
    }
    const pieces = parsed.pathname.replace(/^\/+/, "").split("/");
    const schemeToken = pieces[0];
    const token = pieces[1];
    if (!schemeToken || !token) {
      throw new CodecError("INVALID_URL");
    }
    const match = SCHEME_RE.exec(schemeToken);
    if (!match?.groups?.scheme) {
      throw new CodecError("INVALID_URL");
    }
    const scheme = match.groups.scheme;
    const port = match.groups.port;
    const host = this.decryptHost(token);
    const rest = pieces.length > 2 ? `/${pieces.slice(2).join("/")}` : "/";
    const netloc = port ? `${host}:${port}` : host;
    const query = parsed.search;
    const fragment = parsed.hash;
    return `${scheme}://${netloc}${rest}${query}${fragment}`;
  }
}

export function gatewayHost(gatewayBase: string): string {
  try {
    const url = new URL(gatewayBase);
    if ((url.protocol !== "https:" && url.protocol !== "http:") || !url.hostname) {
      throw new CodecError("INVALID_URL");
    }
    return url.hostname;
  } catch (error) {
    if (error instanceof CodecError) {
      throw error;
    }
    throw new CodecError("INVALID_URL");
  }
}

function keyBytes(value: string | Uint8Array, code: CodecErrorCode): Uint8Array {
  const raw = typeof value === "string" ? utf8(value) : value;
  if (raw.length !== 16) {
    throw new CodecError(code);
  }
  return raw;
}

function utf8(value: string): Uint8Array {
  return new TextEncoder().encode(value);
}

function utf8Decode(value: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(value);
}

function parseUrl(value: string): URL {
  try {
    return new URL(value);
  } catch {
    throw new CodecError("INVALID_URL");
  }
}

function toHex(bytes: Uint8Array): string {
  return [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function fromHex(value: string): Uint8Array | null {
  if (value.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(value)) {
    return null;
  }
  const out = new Uint8Array(value.length / 2);
  for (let i = 0; i < out.length; i += 1) {
    out[i] = Number.parseInt(value.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}
